proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,473 @@
+contains an ``IORef`` is something like::
+
+  data MutPair a b =
+    MutPair {-# UNPACK #-} !(IORef a)
+            {-# UNPACK #-} !(IORef b)
+
+and the runtime representation of this is 3 heap objects: the outer
+``MutPair`` with 2 pointer fields (3 words total), and 2 ``MutVar#``
+objects (2 words each).
+
+Under this proposal, we would be able to declare a ``MutPair`` that is
+represented by a single 3-word heap object containing 2 directly
+mutable pointer fields.
+
+Removing a layer of indirection can make a big difference.  Currently
+``IORef`` consists of two objects: the outer ``IORef`` constructor and
+the inner ``MutVar#`` type.  This proposal allows the two layers to be
+collapsed. Operations on the collapsed ``IORef`` are approximately 2x
+faster than the current ``IORef`` (according to `rough
+experiments <https://gist.github.com/simonmar/d8a05797c01799abcb979aacf27164c7>`_).
+Note that the outer constructor in the current ``IORef``
+representation can be eliminated by using ``UNPACK`` or
+sometimes automatically by strictness analysis, but not always.  The proposed
+representation is *always* cheap.
+
+This extension is aimed at particularly performance-critical data
+structures, where the library writer is prepared to invest some effort
+into defining an optimal representation.  This extension makes it
+possible to eliminate a layer of indirection in mutable data
+structures in a type-safe way, where currently this can only be done
+by using ``unsafeCoerce`` tricks; see for example `the structs
+package <http://hackage.haskell.org/package/structs>`_.
+
+As a special case, this proposal also provides for the declaration of
+constructors with *identity*. That is, constructors that are created
+by an explicitly effectful operation, and that can be compared using
+pointer-equality.  Currently only a few built-in types like ``IORef``
+have identity, so we can currently simulate identity using an
+``IORef`` proxy, but this proposal allows constructors with identity
+to be defined natively.
+
+Proposed Change
+---------------
+
+We propose the addition of a new extension::
+
+  {-# LANGUAGE MutableFields #-}
+
+which enables constructor fields to be declared to be directly
+mutable. A constructor with mutable fields is defined using GADT
+syntax::
+
+  data MutPair a b where
+    MutPair :: mutable a -> mutable b -> IO (MutPair a b)
+
+Note:
+
+- The new ``mutable`` keyword declares a mutable field
+- The return type is in ``IO``: a constructor with any ``mutable``
+  fields *must* have a return type that has one of the forms ``IO t``,
+  ``ST s t``, or ``State# s -> (# State# s, t #)``, where ``t`` takes
+  the form of the normal return type for the constructor. In the
+  latter two cases, the type variable ``s`` must appear in ``t``.
+- mutable fields *must not* have a strictness annotation. (we
+  anticipate that support for strictness annotations on mutable fields
+  will be a future proposal).
```

Comment:
I recommend having the default be strict, as that is more likely to be what people who care about performance actually want.